pybikes/gobike.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2015, Ben Caller <bcaller@gmail.com>
+# Distributed under the LGPL license, see LICENSE.txt
+
+import json
+
+from lxml import html
+from lxml.cssselect import CSSSelector
+
+from .base import BikeShareSystem, BikeShareStation
+from . import utils
+from contrib import TSTCache
+
+# The number of free slots is unavailable
+# However, you can lock the bike next to any full station
+
+PAGE_SIZE = 10
+STATION_LIST_PATH = '/umbraco/Surface/BookingSurface/GetStationsList'
+
+
+class GoBike(BikeShareSystem):
+    sync = True
+
+    meta = {
+        'system': 'GoBike',
+        'company': 'Gobike A/S'
+    }
+
+    def __init__(self, tag, meta, url, availability_path):
+        super(GoBike, self).__init__(tag, meta)
+        self.url = url
+        self.availability_path = availability_path
+
+    def update(self, scraper = None):
+        scraper = scraper or utils.PyBikesScraper()
+
+        # Request station names and locations from JSON file
+        stations = json.loads(scraper.request(self.url + STATION_LIST_PATH))["List"]
+        stations_by_id = {str(s["UnifiedId"]): GoBikeStation(s) for s in stations}
+
+        # Request number of bikes, at 10 stations per page
+        for page in self._get_all_pages(scraper, len(stations)):
+            for uid, bikes in self._parse_page(page):
+                stations_by_id[uid].bikes = bikes
+
+        self.stations = stations_by_id.values()
+
+    def _get_all_pages(self, scraper, n_stations):
+        n_pages = n_stations/PAGE_SIZE + (n_stations % PAGE_SIZE > 0)
+        for p in range(0, n_pages):
+            data = {
+                'lat': self.meta['latitude'],
+                'lng': self.meta['longitude'],
+                'page': p
+            }
+            yield scraper.request(self.url + self.availability_path, 'POST', data=data)
```

Comment:
for instance, this append is done for each page, when it could have been constructed on __init__, since the URL is always the same